---
title: "Working with the Tidyverse"
output: html_notebook
---

We're going to practice using some of the functions in the `dplyr` library. First, lets install the `tidyverse` packages
```{r}
install.packages('tidyverse')
```

You will be using the pipe function all lot; I recommend setting a key board short cut for it. Go to Tools > modify keyboard searchcuts > search for "pipe"  >  click to change the command. I find `alt + 2` or `option + 2` work well.

load the `dplyr` library
```{r}

```

we'll be working with the `starwars` dataset

```{r}
data("starwars")
starwars
```

Select the `Name` column from the `starwars` dataset. Store this in a variable 

```{r}

```

now `pull` the `Name` column from the dataset. store this in a variable

```{r}

```

Notice the differences in the returned data. use `class` to examine both outputs 

```{r}

```

select the name and height columns 

```{r}

```

 Using pipes to add to the previous command,  add a column with the height in inches.  the height column is currently in cm and  1 cm is 0.393701 inches 
```{r}

```


## Example - GTF file

- a GTF file contains information about the location of genes within the genome, as well as vartiety of other annotations. `../src/gtf_example.tsv.gz` contains 5 genes. 
- Each gene has multiple transcripts, and each transcript has a protein coding region, or CDS.  read this data in, and use functions from the tidyverse to calculate the length of each CDS, and determine the gene with the longest CDS. Generaly workflow  
read the data  >  select values that only correspond the to CDS(look at the `type` column)  >  for each CDS group together and calculate the smallest start and the largest end( the `transcript_id` column is unique for each CDS ) >  largest end - smallest start  = length

```{r}

```

## Tying it all together

- the file  "../src/tx_expression_subset.tsv.gz" has some data from a real RNA-seq experiment and contains transcript expression values; there is a column fortranscript name, and the rest of the columns are the samples each transcript is expressed in. 
- Using the gene to transcript id mapping present in the gtf file from above, add gene names to this expression data, and select the genes DCX , GRM3 , MYO7A. summarise the transcript level expression to gene level, and make box plots for  the expression of each gene expression. 

```{r}

```
 










